streamer/m3u8_concater.py

```diff
@@ -0,0 +1,957 @@
+      """Return the elements of the list which are not None."""
+      
+      return [item for item in items if item is not None]
+    
+    # Extract all the languages from every stream.
+    langs: Set[str] = set()
+    for txt_playlists in all_txt_playlists:
+      for txt_playlist in txt_playlists:
+        langs.add(txt_playlist.language)
+    
+    # Create and initialize a division map to divide streams of different
+    # languages into.
+    division: Dict[str, List[Optional[MediaPlaylist]]] = {}
+    for lang in langs:
+      division[lang] = []
+      # Assume all media playlists for lang are not there at the begining.
+      for _ in range(len(all_txt_playlists)):
+        # None = no media playlist with this language for this period.
+        division[lang].append(None)
+    
+    # Add the ith media playlist in the ith position for its language division.
+    for i, txt_playlists in enumerate(all_txt_playlists):
+      for txt_playlist in txt_playlists:
+        # Assume that for any langauge, there is at most one stream for it.
+        # NOTE: A duplicate second language will overwrite the first one.
+        division[txt_playlist.language][i] = txt_playlist
+    
+    # Try to substitute for missing languages if possible.
+    for i in range(len(all_txt_playlists)):
+      for lang in langs:
+        if not division[lang][i]:
+          # Get all the text playlists persent in this period.
+          txt_playlist_options = non_nones([division[lg][i] for lg in langs])
+          sub_lang = MediaPlaylist._fit_missing_lang(txt_playlist_options,
+                                                     lang)
+          division[lang][i] = division[sub_lang][i]
+    
+    concat_txt_playlists: List[MediaPlaylist] = []
+    for lang, optional_txt_playlists in division.items():
+      stream_info = MediaPlaylist._similar_stream_info(
+          # There must be at least one that isn't None.
+          non_nones(optional_txt_playlists))
+      if lang != 'und':
+        # Put the language attribute in case it was removed.
+        stream_info['LANGUAGE'] = _quote(lang)
+      # Get a unique name and uri.
+      stream_info['NAME'], stream_info['URI'] = MediaPlaylist._next_unq_name()
+      concat_txt_playlist = MediaPlaylist(stream_info)
+      # Set the target duration of the concat playlist to the max of 
+      # all children playlists.
+      concat_txt_playlist.target_duration = MediaPlaylist._max_targer_dur(
+          non_nones(optional_txt_playlists))
+      for i, optional_txt_playlist in enumerate(optional_txt_playlists):
+        if optional_txt_playlist:
+          # If a playlist is there, append it.
+          concat_txt_playlist.content += optional_txt_playlist.content
+        else:
+          # If the no playlist were found for this period,
+          # Create a time gap filled with dummy data for the period's duration.
+          dummy = ',\ndata:text/plain,NO {} SUBTITLES\n'.format(lang.upper())
```

Comment:
I haven't actually tested whether this renders on the screen normally or not, i have a problem with subtitles in HLS, their streams are being dropped in chrome and firefox too. Will try to fix it and report back.

Reply:
I think it's better to leave it a blank string as you stated. Thanks!

> It's weird, though, and maybe hacky, to put the comma and newline before the URI as part of this string. That would more logically belong to the EXTINF string.

Was fighting with line lengths T_T